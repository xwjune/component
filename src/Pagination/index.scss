.pagination {
  margin: 12px 20px;
  list-style: none;
  box-sizing: border-box;
  text-align: right;
  > li {
    padding: 0;
    margin: 0 0 0 8px;
    display: inline-block;
    min-width: 32px;
    height: 32px;
    line-height: 32px;
    text-align: center;
    list-style: none;
    color: rgba(0, 0, 0, 0.65);
    transition: all .3s;
    input,
    select {
      display: inline-block;
      padding: 4px 11px;
      width: 100%;
      height: 32px;
      font-size: 14px;
      line-height: 1.5;
      color: rgba(0, 0, 0, 0.65);
      background-color: #fff;
      background-image: none;
      border: 1px solid #d9d9d9;
      border-radius: 4px;
      transition: all .3s;
      margin: 0 4px;
      width: 50px;
      outline: 0;
    }
    select {
      width: auto;
    }
    input:hover,
    input:focus {
      border-color: #409aff;
    }
    div + div {
      margin-left: 8px;
    }
  }
  .options {
    margin-right: 16px;
  }
  .prev,
  .next {
    color: rgba(0, 0, 0, .65);
    user-select: none;
    cursor: pointer;
  }
  .pages,
  .jumper {
    display: inline-block;
  }
  .disabled,
  .disabled:hover,
  .disabled:focus {
    cursor: not-allowed;
    color: rgba(0, 0, 0, 0.25) !important;
  }
}
