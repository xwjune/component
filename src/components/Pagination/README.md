> 缺省总条数的分页

## API

| Name | Description | Type | Default |
| :--- | :--- | :--- | :--- | :--- |
| prefixCls | 样式前缀 | string | jun-pagination |
| className | 样式类 | string  | - |
| current | 当前页数 | number | 1 |
| pageSize | 每页条数 | number | 10 |
| dataSize | 当前页数据总数 | number | - |
| onChange | 页码改变的回调，参数是改变后的页码及每页条数 | function(page, pageSize) | noop |
| disabled | 禁用分页 | boolean | false |
| showQuickJumper | 是否可以快速跳转至某页 | boolean | false |
| showSizeChanger | 是否展示`pageSize`切换器 | boolean | false |
| pageSizeOptions | 指定每页可以显示多少条 | string[] | ['10', '20', '30', '40'] |

```jsx
<Pagination
  dataSize={50}
  current={1}
  pageSize={10}
  onChange={this.onPageChange}
/>
```
