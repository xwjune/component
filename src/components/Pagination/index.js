/**
 * 缺省总条数的分页
 * 若未传属性current、pageSize，则组件内部维护
 *
 * @param {number} dataSize - 当前返回数据条数【用于判断是否为最后一页】
 * @param {number} [current] - 当前页码
 * @param {function} [onChange] - 页码改变的回调【function(page, pageSize)】
 * @param {number} [pageSize=10] - 每页条数
 * @param {boolean} [showQuickJumper=false] - 是否可以快速跳转至某页
 * @param {boolean} [showSizeChanger=false] - 是否可以改变pageSize
 * @param {string[]} [pageSizeOptions=['10','20','30','40']] - 指定每页可以显示多少条
 *
 * @date 2018/04/17
 */
import React from 'react';
import PropTypes from 'prop-types';
import cs from 'classnames';
import getPrefixCls from '../../utils/getPrefixCls';

const PAGE_SIZE = 10;

export default class Pagination extends React.PureComponent {
  static propTypes = {
    dataSize: PropTypes.number.isRequired,
    current: PropTypes.number,
    pageSize: PropTypes.number,
    onChange: PropTypes.func,
    showQuickJumper: PropTypes.bool,
    showSizeChanger: PropTypes.bool,
    pageSizeOptions: PropTypes.arrayOf(PropTypes.string),
  };

  static defaultProps = {
    showQuickJumper: false,
    showSizeChanger: false,
    pageSizeOptions: ['10', '20', '30', '40'], // 默认指定每页可以显示条数
  };

  state = {
    pageSizeOptions: this.props.pageSizeOptions, // 每页可以显示条数
    pageSize: PAGE_SIZE, // 每页条数
    current: 1, // 当前页数
    jumpPage: '', // 跳转至指定某页
  };

  componentDidMount() {
    const { pageSize, pageSizeOptions } = this.props;
    if (typeof pageSize !== 'undefined') {
      const set = new Set([...pageSizeOptions, `${pageSize}`]);
      this.setState({
        pageSizeOptions: Array.from(set).sort((a, b) => a - b),
      });
    }
  }

  // 跳转至某页
  onJumper = (e) => {
    if (e.keyCode === 13) {
      // Enter键
      const { onChange } = this.props;
      const { jumpPage } = this.state;
      const pageSize = typeof this.props.pageSize !== 'undefined' ? this.props.pageSize : this.state.pageSize;
      this.setState({ jumpPage: '' });
      if (/^\d+$/.test(jumpPage)) {
        // 页码校验
        const page = Number(jumpPage) || 1;
        if (typeof this.props.current === 'undefined') {
          this.setState({
            current: page,
          });
        }
        if (onChange) {
          onChange(page, pageSize);
        }
      }
    }
  };

  onPageSizeChange = (e) => {
    const { onChange } = this.props;
    const pageSize = Number(e.target.value);
    const current = 1;

    if (typeof this.props.pageSize === 'undefined') {
      this.setState({ pageSize });
    }
    if (typeof this.props.current === 'undefined') {
      this.setState({ current });
    }
    if (onChange) {
      onChange(current, pageSize);
    }
  };

  onJumperChange = (e) => {
    this.setState({
      jumpPage: e.target.value,
    });
  };

  onPageChange(type) {
    const { current, dataSize, onChange } = this.props;
    const pageSize = typeof this.props.pageSize !== 'undefined' ? this.props.pageSize : this.state.pageSize;
    let page = typeof current !== 'undefined' ? current : this.state.current;

    if (type === 'prev') {
      // 前一页
      if (page === 1) return; // 第一页
      if (page > 1) page -= 1;
    } else {
      // 后一页
      if (dataSize < pageSize) return; // 最后一页
      page += 1;
    }

    if (typeof current === 'undefined') {
      this.setState({ current: page });
    }
    if (onChange) {
      onChange(page, pageSize);
    }
  }

  render() {
    const {
      dataSize,
      current,
      showQuickJumper,
      showSizeChanger,
    } = this.props;
    const { pageSizeOptions } = this.state;
    const pageSize = typeof this.props.pageSize !== 'undefined' ? this.props.pageSize : this.state.pageSize;
    const page = typeof current !== 'undefined' ? current : this.state.current;
    const isFirst = page === 1; // 是否第一页
    const isLast = dataSize < pageSize; // 是否最后一页

    return (
      <ul
        className={getPrefixCls('pagination')}
        unselectable="unselectable"
      >
        {(showQuickJumper || showSizeChanger) && (
          <li className="options">
            {showQuickJumper && (
              <div className="jumper">
                跳至
                <input
                  type="text"
                  value={this.state.jumpPage}
                  onChange={this.onJumperChange}
                  onKeyDown={this.onJumper}
                />
                页
              </div>
            )}
            {showSizeChanger && (
              <div className="pages">
                <select
                  value={pageSize}
                  onChange={this.onPageSizeChange}
                >
                  {
                    pageSizeOptions.map((el) => (
                      <option value={el} key={el}>
                        {el}
                        条/页
                      </option>
                    ))
                  }
                </select>
              </div>
            )}
          </li>
        )}
        <li className="pages">
          页码:
          {' '}
          {page}
        </li>
        <li
          className={cs('prev', isFirst ? 'disabled' : '')}
          onClick={() => this.onPageChange('prev')}
        >
          上一页
        </li>
        <li
          className={cs('next', isLast ? 'disabled' : '')}
          onClick={() => this.onPageChange('next')}
        >
          下一页
        </li>
      </ul>
    );
  }
}
