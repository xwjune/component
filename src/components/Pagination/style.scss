@import '../../style/variables';

$pagination-prefix-cls: '#{$cls-prefix}-pagination';

.#{$pagination-prefix-cls} {
  margin: 12px 20px;
  list-style: none;
  box-sizing: border-box;
  text-align: right;
  > li {
    padding: 0;
    margin: 0 0 0 8px;
    display: inline-block;
    min-width: 32px;
    height: 32px;
    line-height: 32px;
    text-align: center;
    list-style: none;
    color: rgba(0, 0, 0, 0.65);
    transition: all .3s;
    input,
    select {
      display: inline-block;
      padding: 4px 11px;
      margin: 0 4px;
      width: 50px;
      height: 32px;
      font-size: 14px;
      line-height: 1.5;
      color: rgba(0, 0, 0, 0.65);
      background-color: #fff;
      background-image: none;
      border: 1px solid #d9d9d9;
      border-radius: 4px;
      transition: all .3s;
      outline: 0;
      box-sizing: border-box;
    }
    select {
      width: auto;
    }
    input {
      &:hover,
      &:focus {
        border-color: #409aff;
      }
    }
  }
  &-prev,
  &-next {
    color: rgba(0, 0, 0, .65);
    user-select: none;
    cursor: pointer;
  }
  &-next + &-options,
  &-next + &-jumper {
    margin-left: 16px;
  }
  &-disabled {
    &,
    &:hover,
    &:focus {
      cursor: not-allowed;
      color: rgba(0, 0, 0, 0.25) !important;
    }
  }
}
