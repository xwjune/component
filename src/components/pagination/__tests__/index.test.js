import React from 'react';
import { shallow, mount } from 'enzyme';
import Pagination from '../index';

const PAGE_SIZE = 10;

describe('Pagination', () => {
  test('renders: disable prev', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={10}
        current={1}
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('renders: disable next', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={8}
        current={2}
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('renders: disable prev and next', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={8}
        current={1}
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('renders: showQuickJumper, showSizeChanger', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={10}
        current={1}
        showQuickJumper
        showSizeChanger
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('click prev and next', () => {
    let wrapper = null;
    const total = 19;
    const onChange = (pageNum, pageSize) => {
      const left = total - (pageNum - 1) * PAGE_SIZE;
      const dataSize = left > PAGE_SIZE ? PAGE_SIZE : left;
      wrapper.setProps({
        dataSize,
        pageSize,
        current: pageNum,
      });
    };
    wrapper = mount(
      <Pagination
        dataSize={10}
        current={1}
        onChange={onChange}
        showQuickJumper
        showSizeChanger
      />
    );
    expect(wrapper.find('.page').text()).toBe('页码: 1');
    // 第一页点击上一页-禁止点击
    wrapper.find('.prev').simulate('click');
    expect(wrapper.prop('current')).toBe(1);

    // 下一页点击
    wrapper.find('.next').simulate('click');
    expect(wrapper.prop('current')).toBe(2);
    expect(wrapper.find('.next').hasClass('disabled')).toBeTruthy();

    // 最后一页点击下一页-禁止点击
    wrapper.find('.next').simulate('click');
    expect(wrapper.prop('current')).toBe(2);

    // 上一页点击
    wrapper.find('.prev').simulate('click');
    expect(wrapper.prop('current')).toBe(1);
    expect(wrapper.find('.prev').hasClass('disabled')).toBeTruthy();

    // 跳转至某页
    wrapper.find('input').simulate('change', {
      target: {
        value: 2,
      },
    });
    wrapper.find('input').simulate('keydown', {
      key: 'Enter',
      keyCode: 13,
    });
    expect(wrapper.prop('current')).toBe(2);

    // 更改每页条数
    wrapper.find('select').simulate('change', {
      target: {
        value: 20,
      },
    });
    expect(wrapper.prop('pageSize')).toBe(20);
  });

  test('current、pageSize内部维护', () => {
    const wrapper = shallow(
      <Pagination dataSize={10} />
    );
    expect(wrapper.state('current')).toBe(1);

    // 下一页点击
    wrapper.find('.next').simulate('click');
    expect(wrapper.state('current')).toBe(2);

    // 上一页点击
    wrapper.find('.prev').simulate('click');
    expect(wrapper.state('current')).toBe(1);
    expect(wrapper.find('.prev').hasClass('disabled')).toBeTruthy();
  });

  test('跳转至某页', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={10}
        showQuickJumper
      />
    );
    // 跳转页码格式错误
    wrapper.find('input').simulate('change', {
      target: {
        value: 3.2,
      },
    });
    wrapper.find('input').simulate('keydown', {
      key: 'Enter',
      keyCode: 13,
    });
    expect(wrapper.state('current')).toBe(1);

    // 正确跳转
    wrapper.find('input').simulate('change', {
      target: {
        value: 2,
      },
    });
    // 非enter键
    wrapper.find('input').simulate('keydown', {
      key: 'Shift',
      keyCode: 16,
    });
    expect(wrapper.state('current')).toBe(1);
    // enter键
    wrapper.find('input').simulate('keydown', {
      key: 'Enter',
      keyCode: 13,
    });
    expect(wrapper.state('current')).toBe(2);
  });

  test('更改每页条数', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={10}
        showSizeChanger
      />
    );
    wrapper.find('select').simulate('change', {
      target: {
        value: 20,
      },
    });
    expect(wrapper.state('pageSize')).toBe(20);
  });

  test('设置pageSizeOptions', () => {
    const wrapper = shallow(
      <Pagination
        dataSize={10}
        pageSize={8}
        pageSizeOptions={['10', '20']}
        showSizeChanger
      />
    );
    expect(wrapper.state('pageSizeOptions')).toEqual(['8', '10', '20']);
  });
});
