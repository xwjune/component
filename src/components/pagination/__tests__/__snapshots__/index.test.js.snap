// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders: disable next 1`] = `
<ul
  class="jun-pagination"
  unselectable="unselectable"
>
  <li
    class="page"
  >
    页码: 2
  </li>
  <li
    class="prev"
  >
    上一页
  </li>
  <li
    class="next disabled"
  >
    下一页
  </li>
</ul>
`;

exports[`Pagination renders: disable prev 1`] = `
<ul
  class="jun-pagination"
  unselectable="unselectable"
>
  <li
    class="page"
  >
    页码: 1
  </li>
  <li
    class="prev disabled"
  >
    上一页
  </li>
  <li
    class="next"
  >
    下一页
  </li>
</ul>
`;

exports[`Pagination renders: disable prev and next 1`] = `
<ul
  class="jun-pagination"
  unselectable="unselectable"
>
  <li
    class="page"
  >
    页码: 1
  </li>
  <li
    class="prev disabled"
  >
    上一页
  </li>
  <li
    class="next disabled"
  >
    下一页
  </li>
</ul>
`;

exports[`Pagination renders: showQuickJumper, showSizeChanger 1`] = `
<ul
  class="jun-pagination"
  unselectable="unselectable"
>
  <li
    class="options"
  >
    <div
      class="jumper"
    >
      跳至
      <input
        type="text"
        value=""
      />
      页
    </div>
    <div
      class="pageSize"
    >
      <select>
        <option
          selected=""
          value="10"
        >
          10条/页
        </option>
        <option
          value="20"
        >
          20条/页
        </option>
        <option
          value="30"
        >
          30条/页
        </option>
        <option
          value="40"
        >
          40条/页
        </option>
      </select>
    </div>
  </li>
  <li
    class="page"
  >
    页码: 1
  </li>
  <li
    class="prev disabled"
  >
    上一页
  </li>
  <li
    class="next"
  >
    下一页
  </li>
</ul>
`;
